IdentityCore/src/oauth2/MSIDTokenResponseHandler.m

```diff
@@ -0,0 +1,172 @@
+#import "MSIDAADV2TokenResponse.h"
+#import "MSIDError.h"
+#import "MSIDClientInfo.h"
+#import "MSIDAADV2RequestParameters.h"
+#import "NSOrderedSet+MSIDExtensions.h"
+
+@implementation MSIDTokenResponseHandler
+
++ (BOOL)processResponse:(MSIDAADTokenResponse *)response
+           refreshToken:(NSString *)refreshToken
+          requestParams:(MSIDRequestParameters *)parameters
+                context:(id<MSIDRequestContext>)context
+                  error:(NSError * __autoreleasing *)error
+{
+    if (!response)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInternal, @"processTokenResponse called without a response dictionary", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    [self.class checkCorrelationId:response requestCorrelationId:context.correlationId];
+    
+    if (response.error)
+    {
+        if (error)
+        {
+            MSIDErrorCode errorCode = refreshToken ? MSIDErrorServerRefreshTokenRejected : MSIDErrorServerOauth;
+            if ([response isKindOfClass:MSIDAADV2TokenResponse.class])
+            {
+                errorCode = [self.class getErrorCodeForAADV2:response.error];
+            }
+            
+            *error = MSIDCreateError(MSIDOAuthErrorDomain,
+                                     errorCode,
+                                     response.errorDescription,
+                                     response.error,
+                                     nil,
+                                     nil,
+                                     context.correlationId,
+                                     nil);
+        }
+        return NO;
+    }
+    
+    if (!response.accessToken)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Authentication response received without expected accessToken", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    if (!response.clientInfo)
+    {
+        MSID_LOG_ERROR(context, @"Client info was not returned in the server response");
+        MSID_LOG_ERROR_PII(context, @"Client info was not returned in the server response");
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInternal, @"Client info was not returned in the server response", nil, nil, nil, context.correlationId, nil);
+        }
+        return NO;
+    }
+    
+    // Checking for V1 response is done
+    if (![response isKindOfClass:MSIDAADV2TokenResponse.class] || ![parameters isKindOfClass:MSIDAADV2RequestParameters.class])
+    {
+        return YES;
+    }
+    
+    // The rest of the checking is only for V2
+    MSIDAADV2RequestParameters *aadV2Parameters = (MSIDAADV2RequestParameters *)parameters;
+    if ([NSString msidIsStringNilOrBlank:response.scope])
```

Comment:
Two comments:
1. this logic is v2 specific, so it shouldn't be here
2. Just curious: what is the case where scope is not returned?

Reply:
As discussed offline, this logic should be moved to token item class.